src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.cs

```diff
@@ -0,0 +1,682 @@
+
+        private delegate T LoadFromFileFunc<T>(
+            ReadOnlyMemory<byte> data,
+            ReadOnlySpan<char> password,
+            X509KeyStorageFlags keyStorageFlags,
+            Pkcs12LoaderLimits loaderLimits);
+
+        private static T LoadFromFile<T>(
+            string path,
+            ReadOnlySpan<char> password,
+            X509KeyStorageFlags keyStorageFlags,
+            Pkcs12LoaderLimits loaderLimits,
+            LoadFromFileFunc<T> loader)
+        {
+            (byte[]? rented, int length, MemoryMappedFile? mapped) = ReadAllBytesIfBerSequence(path);
+
+            if (rented is not null)
+            {
+                Debug.Assert(mapped is null);
+
+                try
+                {
+                    return loader(
+                        new ReadOnlyMemory<byte>(rented, 0, length),
+                        password,
+                        keyStorageFlags,
+                        loaderLimits);
+                }
+                finally
+                {
+                    CryptoPool.Return(rented, length);
+                }
+            }
+            else
+            {
+                Debug.Assert(mapped is not null);
+
+                using (mapped)
+                using (MemoryMappedViewAccessor accessor = mapped.CreateViewAccessor(0, 0, MemoryMappedFileAccess.Read))
+                {
+                    unsafe
+                    {
+                        byte* pointer = null;
+
+                        try
+                        {
+                            accessor.SafeMemoryMappedViewHandle.AcquirePointer(ref pointer);
+
+                            using (PointerMemoryManager<byte> manager = new(pointer, length))
+                            {
+                                return loader(
+                                    manager.Memory,
+                                    password,
+                                    keyStorageFlags,
+                                    loaderLimits);
+                            }
+                        }
+                        finally
+                        {
+                            if (pointer != null)
+                            {
+                                accessor.SafeMemoryMappedViewHandle.ReleasePointer();
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // TODO: Get this made into a public method.
+        private static readonly TryReadLength s_tryReadLength = (TryReadLength)typeof(AsnDecoder).GetMethod(
+            nameof(TryReadLength), BindingFlags.Static | BindingFlags.NonPublic)!.CreateDelegate(typeof(TryReadLength));
+
+        private delegate bool TryReadLength(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            out int? length,
+            out int bytesRead);
+
+        private static (byte[]?, int, MemoryMappedFile?) ReadAllBytesIfBerSequence(string path)
```

Comment:
This is "clever". Do we really want to optimize for the case of 1MB PKCS12 docs, or just read the whole thing in as bytes? We do `File.ReadAllBytes` in many PKCS12 paths already.

Reply:
Checking that it's at least plausibly BER-encoded definitely feels a bit virtuous, just so a multi GB video file misclick doesn't trigger an OOM or "the file's too big to be in an array" exception.

The next bit of virtue is using an array pool instead of GC fodder, though there's a reasonable argument that PFXes shouldn't be loaded often enough to care.

And the bit about MemoryMappedFile is just an experiment/discussion item.

Reply:
Sure. A quick BER sniff test makes sense my comment was specifically on memory mapping it in.